modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -8171,22 +8174,38 @@ final void notifyFocusListeners() {
     }
 
     /**
-     * Called when the current node was removed from the scene graph.
-     * If the current node has the focusWithin bit, we also need to clear the focusWithin bits of this
-     * node's parents. Note that a scene graph can have more than a single focused node, for example when
-     * a PopupWindow is used to present a branch of the scene graph. Since we need to preserve multi-level
+     * Called when the current node was removed from or added to the scene graph.
+     * If the current node has the focusWithin bit, we also need to clear and set the focusWithin bits of this
+     * node's old and new parents. Note that a scene graph can have more than a single focused node, for example
+     * when a PopupWindow is used to present a branch of the scene graph. Since we need to preserve multi-level
      * focus, we need to adjust the focus-within count on all parents of the node.
      */
-    private void clearParentsFocusWithin(Node oldParent) {
-        if (oldParent != null && focusWithin.get()) {
-            Node node = oldParent;
-            while (node != null) {
-                node.focusWithin.adjust(-focusWithin.count);
-                node = node.getParent();
-            }
+    private void updateParentsFocusWithin(Node oldParent, Node newParent) {
+        if (!focusWithin.get()) {
+            return;
+        }
 
+        Node node = oldParent;
+        while (node != null) {
+            node.focusWithin.adjust(-focusWithin.count);
+            node = node.getParent();
+        }
+
+        node = newParent;
+        while (node != null) {
+            node.focusWithin.adjust(focusWithin.count);
+            node = node.getParent();
+        };
+
+        // Since focus changes are atomic, we only fire change notifications after
+        // all changes are committed on all old and new parents.
```

Comment:
Is it possible a user can interfere with this process by adding a listener to `focusWithinProperty`?  It fires events when it goes from `true` to `false` and vice versa.

Reply:
I also noticed this code, which I think will not do the correct thing when `change` is not 1 or -1:

        void adjust(int change) {
            count += change;

            if (count == 1) {
                set(true);
            } else if (count == 0) {
                set(false);
            }
        }

For example, if focus within currently is `false`, and it is "adjusted" by 5, the count goes from 0 to 5 in one go.  It will remain `false` then.  Is this intended?

Reply:
> Is it possible a user can interfere with this process by adding a listener to `focusWithinProperty`? It fires events when it goes from `true` to `false` and vice versa.

Can disregard this, it seems the property uses a special base class that breaks with the general property contract.

Reply:
> I also noticed this code, which I think will not do the correct thing when `change` is not 1 or -1:
> 
> ```
>     void adjust(int change) {
>         count += change;
> 
>         if (count == 1) {
>             set(true);
>         } else if (count == 0) {
>             set(false);
>         }
>     }
> ```
> 
> For example, if focus within currently is `false`, and it is "adjusted" by 5, the count goes from 0 to 5 in one go. It will remain `false` then. Is this intended?

Good catch! I've added another test for this edge case.